### PR TITLE
feat: token introspection, api clients and auth method private_key_jwt

### PR DESCRIPTION
**Pull request by livio-a**

### CODE REVIEW

File: internal/api/grpc/management/application_converter.go [hunk 1]
```diff
@@ -432,3 +496,126 @@ func appChangesToMgtAPI(changes *proj_model.ApplicationChanges) (_ []*management
...
+	for i, key := range keys {
+		keyViews[i] = clientKeyViewFromModel(key)
+	}
+	return keyViews
+}
+
+func clientKeyViewFromModel(key *key_model.AuthNKeyView) *management.ClientKeyView {
+	creationDate, err := ptypes.TimestampProto(key.CreationDate)
+	logging.Log("MANAG-DAs2t").OnError(err).Debug("unable to parse timestamp")
+
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/api/grpc/management/application_converter.go:511]: keys can only be created and not changed

File: internal/api/grpc/management/application_converter.go [hunk 2]
```diff
@@ -432,3 +496,126 @@ func appChangesToMgtAPI(changes *proj_model.ApplicationChanges) (_ []*management
...
+}
+
+func addClientKeyToModel(key *management.AddClientKeyRequest) *proj_model.ClientKey {
+	expirationDate := time.Time{}
+	if key.ExpirationDate != nil {
+		var err error
+		expirationDate, err = ptypes.Timestamp(key.ExpirationDate)
+		logging.Log("MANAG-Dgt42").OnError(err).Debug("unable to parse expiration date")
+	}
+
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/api/grpc/management/application_converter.go:531]: keys can only be created and not changed

File: internal/project/repository/view/model/application.go [hunk 3]
```diff
@@ -156,14 +156,20 @@ func (a *ApplicationView) AppendEvent(event *models.Event) (err error) {
 		}
 		a.setCompliance()
 		return a.setOriginAllowList()
-	case es_model.OIDCConfigChanged,
-		es_model.ApplicationChanged:
+	case es_model.APIConfigAdded:
+		a.IsOIDC = false
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/project/repository/view/model/application.go:160]: agree... but should we change it now or in the query refactoring?

File: internal/management/repository/eventsourcing/eventstore/project.go [hunk 4]
```diff
@@ -405,14 +407,59 @@ func (repo *ProjectRepo) ApplicationChanges(ctx context.Context, id string, appI
...
+		Result:      key_view_model.AuthNKeysToModel(keys),
+	}
+	if sequenceErr == nil {
+		result.Sequence = sequence.CurrentSequence
+		result.Timestamp = sequence.LastSuccessfulSpoolerRun
+	}
+	return result, nil
+}
+
+func (repo *ProjectRepo) GetClientKey(ctx context.Context, projectID, applicationID, keyID string) (*key_model.AuthNKeyView, error) {
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/management/repository/eventsourcing/eventstore/project.go:431]: should i check if new events exist?

File: internal/project/repository/view/model/application.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgerschwiler) [reply internal/project/repository/view/model/application.go:160]: we can do it in query refactoring

File: internal/management/repository/eventsourcing/eventstore/project.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgerschwiler) [reply internal/management/repository/eventsourcing/eventstore/project.go:431]: yes

**Review by fgerschwiler — APPROVED** [review]

**Pull request merged**